ls1intum/Artemis
# Feature/achievements/checks

### Checklist

no client changes included

### Description
These changes add the functionality to check whether a student has earned an achievement.
This is done by using an `EntityListener` for the results. Every time a new result is saved or an existing result is updated, this `ResultListener` is invoked and calls the `checkForAchievements()` function from the `AchievementService` with the newly created/modified result as parameter. If the user has earned an achievement, it is assigned to him.

### Steps for Testing
Since there is no client side yet which allows us to check whether we got an achievement or not the only testing needed is to make sure no existing functionality is broken. Especially when generating new results since then the `ResultListener` is triggered and executes the new code.
Aside from that code reviews would be helpful.

### Test Coverage
* ProgressBasedAchievementService.java: 89%
* AchievementService.java 97%
* TimeBasedAchievementService.java 97%
* PointBasedAchievementService.java 97%
* AchievementResource.java 100%
